src/chatd.cpp

```diff
@@ -2667,28 +2684,121 @@ void Chat::sendSync()
+        if (auxReaction.mMsgId == msgId
+            && auxReaction.mReactionString == reaction)
+        {
+            auxReaction.mStatus = status;
             return;
+        }
+    }
 
-        mCrypto->reactionEncrypt(message, reaction)
-        .then([this, wptr, &message](std::shared_ptr<Buffer> data)
+    mPendingReactions.emplace_back(reaction, encReaction, msgId, status);
+}
+
+void Chat::removePendingReaction(const string &reaction, Id msgId)
+{
+    for (auto it = mPendingReactions.begin(); it != mPendingReactions.end(); it++)
+    {
+        if (it->mMsgId == msgId
+            && it->mReactionString == reaction)
         {
-            if (wptr.deleted())
-                return;
+            mPendingReactions.erase(it);
+            return;
+        }
+    }
+}
 
-           std::string encReaction (data->buf(), data->bufSize());  // lenght must be only 1 byte. passing the buffer uses 4 bytes for size
-           sendCommand(Command(OP_ADDREACTION) + mChatId + client().myHandle() + message.id() + (int8_t)data->bufSize() + encReaction);
-        })
-        .fail([this](const ::promise::Error& err)
+void Chat::retryPendingReactions()
+{
+    for (auto& reaction: mPendingReactions)
+    {
+        assert(!reaction.mReactionStringEnc.empty());
+        sendCommand(Command(reaction.mStatus) + mChatId + client().myHandle() + reaction.mMsgId + (int8_t)reaction.mReactionStringEnc.size() + reaction.mReactionStringEnc);
+    }
+}
+
+void Chat::cleanPendingReactionsByMsg(karere::Id msgid, Idx idx)
+{
+    for (auto it = mPendingReactions.begin(); it != mPendingReactions.end();)
+    {
+        auto auxit = it++;
+        if (idx != CHATD_IDX_INVALID)
         {
-            CHATID_LOG_DEBUG("Error encrypting reaction: %s", err.what());
-        });
+            Idx currentIdx = msgIndexFromId(auxit->mMsgId);
+            assert(currentIdx != CHATD_IDX_INVALID);
+            if (currentIdx <= idx)
+            {
+                mPendingReactions.erase(auxit);
+            }
+        }
+        else if (auxit->mMsgId == msgid)
+        {
+            mPendingReactions.erase(auxit);
+        }
+    }
+}
+
+void Chat::flushChatPendingReactions()
+{
+    for (auto it = mPendingReactions.begin(); it != mPendingReactions.end();)
+    {
+        auto auxit = it++;
+        karere::Id msgid = auxit->mMsgId;
+        Idx index = msgIndexFromId(msgid);
+        if (index == CHATD_IDX_INVALID)
+        {
+            // couldn't find msg
+            CHATID_LOG_WARNING("flushChatPendingReactions: message with id(%d) not loaded in RAM", msgid);
+        }
+        else
+        {
+            Message &message = at(index);
+            CALL_LISTENER(onReactionUpdate, message.mId, auxit->mReactionString.c_str(), message.getReactionCount(auxit->mReactionString.c_str()));
+        }
+        CALL_DB(cleanPendingReactions, auxit->mMsgId);
+        mPendingReactions.erase(auxit);
```

Comment:
Instead of erasing item by item, you can iterate as usual, with a `for (auto &reaction : mPendingReactions)` and after the loop, do a `mPendingReactions.clear()`

Reply:
Fixed